### PR TITLE
Upgrade backend to Next.js 15

### PR DESCRIPTION
**Pull request by N2D4**
> [!IMPORTANT]
> Upgrade backend to Next.js 15, update dependencies, modify workflows, and adjust code for compatibility.
> 
>   - **Dependencies**:
>     - Upgrade `next` to `15.0.3` in `apps/backend/package.json` and `examples/demo/package.json`.
>     - Update `react` and `react-dom` to `^19.0.0` in `apps/backend/package.json`.
>     - Update `@types/react` to `link:@types/react@18.3.12` in multiple `package.json` files.
>   - **Workflows**:
>     - Add `.github/workflows/check-codegen.yaml` to ensure codegen is up-to-date.
>     - Remove codegen step from `lint-and-build.yaml`.
>   - **Code Changes**:
>     - Modify `authorize/[provider_id]/route.tsx` and `callback/[provider_id]/route.tsx` to use `await cookies()`.
>     - Change `createSmartRequest` in `smart-request.tsx` to handle `params` as a `Promise`.
>     - Update logging in `smart-route-handler.tsx` to log `statusError.stack` instead of `statusError`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=stack-auth%2Fstack&utm_source=github&utm_medium=referral)<sup> for 61b4009ee185adabd3a473b4c43911ea32cc719d. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: apps/backend/src/middleware.tsx [hunk 1]
```diff
@@ -54,8 +54,7 @@ export async function middleware(request: NextRequest) {
   const isApiRequest = url.pathname.startsWith('/api/');
 
   const newRequestHeaders = new Headers(request.headers);
-  // store the direct IP address of the requester or proxy so we can read it with `headers()` later
-  newRequestHeaders.set("x-stack-direct-requester-or-proxy-ip", request.ip ?? '');
```
Comment (fomalhautb) [inline apps/backend/src/middleware.tsx:58]: What was this for and why don't we need it anymore?

File: dependencies.compose.yaml [hunk 2]
```diff
@@ -58,14 +58,16 @@ services:
   svix-redis:
     image: docker.io/redis:7-alpine
     command: --save 60 500 --appendonly yes --appendfsync everysec --requirepass PASSWORD-PLACEHOLDER--oVn8GSD6b9
+    ports:
```
Comment (fomalhautb) [inline dependencies.compose.yaml]: why do we expose redis now?

**Review by fomalhautb — APPROVED** [review]
looks good

File: apps/backend/src/middleware.tsx (hunk 1 above)
Comment (N2D4) [reply apps/backend/src/middleware.tsx:58]: It was for the geolocation stuff, Next 15 removed `request.ip` so now we just don't check it anymore (we still check various header as a fallback)

File: dependencies.compose.yaml (hunk 2 above)
Comment (N2D4) [reply dependencies.compose.yaml]: mb, good catch

**Pull request merged**